src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+
+    for (uint32_t d = 0u; d < depth; d++)
+    {
+        seed += seeds [d % 10];
+        // TODO add assert that str is not NULL
+        MurmurHash3_x64_128 (str, strlen (str), seed, hash);
+        uint32_t bin = (hash[0] ^ hash[1] ^ hash[2] ^ hash[3]) % width;
+        n = snprintf (path_key+cx, len-cx, "%x.", bin);
+        cx += n;
+        if (cx >= len || n < 0) {
+            return -1;
+        }
+    }
+    n = snprintf (path_key+cx, len-cx, "%s", str);
+    if (cx + n >= len || n < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+int dyad_init(bool debug, bool check, bool shared_storage,
+        unsigned int key_depth, unsigned int key_bins,
+        const char *kvs_namespace, const char *prod_managed_path,
+        const char *cons_managed_path, bool intercept, dyad_ctx_t **ctx)
+{
+    printf("In dyad_init\n");
+    if (*ctx != NULL)
+    {
+        if ((*ctx)->initialized)
+        {
+            // TODO Indicate already initialized
+            DPRINTF ((*ctx), "DYAD context already initialized\n");
+        }
+        else
+        {
+            **ctx = dyad_ctx_default;
+        }
+        return DYAD_OK;
+    }
+    printf("Allocating ctx\n");
+    *ctx = (dyad_ctx_t*) malloc(sizeof(struct dyad_ctx));
+    if (*ctx == NULL)
+    {
+        fprintf (stderr, "Could not allocate DYAD context!\n");
+        return DYAD_NOCTX;
+    }
+    printf("Setting ctx values that don't need allocation\n");
+    **ctx = dyad_ctx_default;
+    (*ctx)->debug = debug;
+    (*ctx)->check = check;
+    (*ctx)->shared_storage = shared_storage;
+    (*ctx)->key_depth = key_depth;
+    (*ctx)->key_bins = key_bins;
+    (*ctx)->intercept = intercept;
+    printf("Setting KVS namespace\n");
+    if (kvs_namespace == NULL)
+    {
+        fprintf (stderr, "No KVS namespace provided!\n");
+        // TODO see if we want a different return val
+        return DYAD_NOCTX;
+    }
+    (*ctx)->kvs_namespace = (char*) malloc(strlen(kvs_namespace)+1);
+    if ((*ctx)->kvs_namespace == NULL)
+    {
+        fprintf (stderr, "Could not allocate buffer for KVS namespace!\n");
+        free(*ctx);
+        *ctx = NULL;
+        return DYAD_NOCTX;
+    }
+    strncpy(
+            (*ctx)->kvs_namespace,
+            kvs_namespace,
+            strlen(kvs_namespace)+1
+           );
+    printf("Setting prod_managed_path\n");
+    if (prod_managed_path == NULL)
+    {
+        (*ctx)->prod_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->prod_managed_path = (char*) malloc(strlen(prod_managed_path)+1);
+        if ((*ctx)->prod_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Producer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->prod_managed_path,
+                prod_managed_path,
+                strlen(prod_managed_path)+1
```

Comment:
There is no error with this way of handling null-termination. However, imagine the case we would need to support string of wide characters, in which case the null-termination would require more than 1 byte. I think it is safer to do null-termination in more explicit fashion either via the character buffer initialization or just writing a null-termination character at the end.